src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs

```diff
@@ -163,14 +163,18 @@ internal static void AdvanceToPosition(this Stream stream, long position)
         {
             long numBytesLeft = position - stream.Position;
             Debug.Assert(numBytesLeft >= 0);
-            while (numBytesLeft != 0)
+            if (numBytesLeft > 0)
             {
                 const int throwAwayBufferSize = 64;
-                int numBytesToSkip = (numBytesLeft > throwAwayBufferSize) ? throwAwayBufferSize : (int)numBytesLeft;
-                int numBytesActuallySkipped = stream.Read(new byte[throwAwayBufferSize], 0, numBytesToSkip);
-                if (numBytesActuallySkipped == 0)
-                    throw new IOException(SR.UnexpectedEndOfStream);
-                numBytesLeft -= numBytesActuallySkipped;
+                byte[] throwAwayBuffer = new byte[throwAwayBufferSize];
```

Comment:
Using the above benchmark (which I'm not sure how real world, or how common having those "Extra fields" is), ArrayPool seems to perform worse:

```C#
        internal static void AdvanceToPosition(this Stream stream, long position)
        {
            long numBytesLeft = position - stream.Position;
            Debug.Assert(numBytesLeft >= 0);
            if (numBytesLeft > 0)
            {
                const int bufferSize = 64;
                byte[] buffer = ArrayPool<byte>.Shared.Rent(bufferSize);

                do
                {
                    int numBytesToSkip = (numBytesLeft > bufferSize) ? bufferSize : (int)numBytesLeft;
                    int numBytesActuallySkipped = stream.Read(buffer, 0, numBytesToSkip);
                    if (numBytesActuallySkipped == 0)
                        throw new IOException(SR.UnexpectedEndOfStream);
                    numBytesLeft -= numBytesActuallySkipped;
                } while (numBytesLeft > 0);

                ArrayPool<byte>.Shared.Return(buffer);
            }
        }
```

|                       Method |              Toolchain |     Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 |  Gen 1 | Allocated |
|----------------------------- |----------------------- |---------:|----------:|----------:|------:|--------:|-------:|-------:|----------:|
| ReadZipContainingExtraFields |      \main\corerun.exe | 1.143 us | 0.0130 us | 0.0116 us |  1.00 |    0.00 | 0.4692 | 0.0038 |      3 KB |
| ReadZipContainingExtraFields |        \pr\corerun.exe | 1.036 us | 0.0202 us | 0.0179 us |  0.91 |    0.02 | 0.2575 | 0.0019 |      2 KB |
| ReadZipContainingExtraFields | \arrayPool\corerun.exe | 1.084 us | 0.0212 us | 0.0253 us |  0.95 |    0.02 | 0.2441 | 0.0019 |      2 KB |

Do you think this matters much?

Reply:
I don't know how frequently this operation is performed.  If it's relatively rare, just allocating is fine.  If it might be more common, I'd go with the ArrayPool version, and if the numBytesLeft might be larger, renting a larger buffer, which should in turn actually make it more of a win.

Reply:
> If it's relatively rare

It appears to be relatively rare, given the unit tests only hit it in these 3 tests:

https://github.com/dotnet/runtime/blob/4db986ef656ab39b0ed93b59c423119abbdcea27/src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs#L812-L861

Modifying the .zip data used in those tests to add more "Extra Fields" is where I got the Benchmark zip from. I honestly questioned making the PR at all, but I felt I needed to since allocating a new buffer each time through the loop seemed really bad.

Reply:
Then just moving it out of the loop seems fine.